src/frames.ts

```diff
@@ -1108,3 +935,146 @@ class PendingNavigationBarrier {
       this._promiseCallback();
   }
 }
+
+export class FrameTask {
+  private _frame: Frame;
+  private _failurePromise: Promise<Error>;
+  private _requestMap = new Map<string, network.Request>();
+  private _disposables: (() => void)[] = [];
+  private _url: string | undefined;
+
+  constructor(frame: Frame, options: types.TimeoutOptions, url?: string) {
+    this._frame = frame;
+    this._url = url;
+
+    // Process timeouts
+    let timeoutPromise = new Promise<TimeoutError>(() => {});
+    const { timeout = frame._page._timeoutSettings.navigationTimeout() } = options;
+    if (timeout) {
+      const errorMessage = 'Navigation timeout of ' + timeout + ' ms exceeded';
+      let timer: NodeJS.Timer;
+      timeoutPromise = new Promise(fulfill => timer = setTimeout(fulfill, timeout))
+          .then(() => { throw new TimeoutError(errorMessage); });
+      this._disposables.push(() => clearTimeout(timer));
+    }
+
+    // Process detached frames
+    this._failurePromise = Promise.race([
+      timeoutPromise,
+      this._frame._page._disconnectedPromise.then(() => { throw new Error('Navigation failed because browser has disconnected!'); }),
+      this._frame._detachedPromise.then(() => { throw new Error('Navigating frame was detached!'); }),
+    ]);
+
+    // Collect requests during the task.
+    const watcher = (request: network.Request) => {
+      if (!request._documentId || request.redirectedFrom())
+        return;
+      this._requestMap.set(request._documentId, request);
+    };
+    this._disposables.push(() => this._frame._requestWatchers.delete(watcher));
+    this._frame._requestWatchers.add(watcher);
+  }
+
+  async raceAgainstFailures<T>(promise: Promise<T>): Promise<T> {
+    let result: T;
+    let error: Error | undefined;
+    await Promise.race([
+      this._failurePromise.catch(e => error = e),
+      promise.then(r => result = r).catch(e => error = e)
+    ]);
+
+    if (!error)
+      return result!;
+    this.done();
```

Comment:
Calling done() at the right time is a bit tricky. Returning error was easier in that regard.